### PR TITLE
Autobounty v1.0

### LINKED ISSUE #1: Make Autobounty Bot Production Ready
We have forked the initial nodejs bot put together by the Aragon team that is intended to automatically fund bounties, rather than having to require one of their team members to fund each and every bounty they issue manually.

The idea is that organizations can run this bot themselves, and maintain a (very) hot wallet that they top up once a week (depending on how heavily they use OpenBounty) rather than funding bounties one by one.

There are a number of requirements for this bounty:

1) Change the bot to look for **label sizes** (`bounty-xs`, `bounty-s`, `bounty-m`, `bounty-l`, `bounty-xl` and `bounty-xxl`) and fund the bounties according to them, rather than looking for a webhook based on comments. This is to ensure that only users with write access to the repo can set the amount to be funded (as only those with write access can assign labels).
2) Hook into ethgasstation.info's API, or the equivalent, in order to fetch dynamic gas prices and ensure that transactions sent from the funding wallet do not fail, or take too long to get picked up. If no response comes back from the gas station, default to the usual `21 GWei`.
3) Set up some configurable ENV variable apart from the wallet keys such that organizations can define roughly the `number of hours` per label (i.e. `bounty-s` = average 3 hrs, `bounty-l` = average 14hrs at Status) and the `price per hour` they are willing to pay (at Status, $35/hr) so that the bounties sent to issues can be done so dynamically.
4) Fetch token price for that organization so that we can convert the above `number of hours` * `price per hr` into a real-time token amount. Currently, we fetch this info for SNT here: `https://min-api.cryptocompare.com/data/price?fsym=SNT&tsyms=USD` - so just make the `SNT` part another ENV variable so that organizations can use any ERC20/ERC223 token of their choice.
5) Make it easy and obvious for organisations to set up the required ENV variables to handle keys and bounty sizes and edit the README to reflect this.
6) As this introduces some security concerns, it is absolutely necessary that there are some solid tests for the majority of the functionality here. This is the major reason that this bounty has a `bounty-xl` tag - we require a good testing suite to merge it as a solution as we need to be sure that it works as expected and will not result in unnecessary loss of funds.

### PR DESCRIPTION
**Pull request by kafkasl**
Fixes #1 

First fully functional version of the autobounty bot.

All implementation and configuration details can be found on the readme.md file.

### CODE REVIEW

**Comment by andytudhope** [comment]
Amazing work, thank you @kafkasl and @jomsdev!! Will do a full review and QA run early next week and hopefully merge ASAP! Thank you 🎉 ❤️

**Comment by churik** [comment]
## 1. I didn't manage to make it work.

[config.txt](https://github.com/status-im/autobounty/files/1677257/config.txt) - used as `production.js`
[Repository](https://github.com/NewOrGanizationSOB/RepoSOB) - repository with configured webhook webh 2
Organization owner in user `Chu127001`

1.1. `sourceAddress` and  `githubUsername` from config should match address from `Payment details` and GH repo owner of particular organization? So, i.e. if I 
GH username - "AAA" is owner of organization "BBB", which is owner of repo "CCC" and I have installed and configured MM account on SOB with address "010101"  my config details should look like:
`githubUsername`: "AAA"
`sourceAddress`: "010101"

1.2. Where is credentials for eth account in your config? how bot should access to '0xcf2272205cc0cf96Cfbb9Dd740BD681D1E86901E' and sign transaction?

1.3. From `readme.md` 

> Configure the webhook to be triggered by comments in issues selecting the Issue Comment box in 'Let me select individual events'


but in fact we add label for bounty (i.e. `bounty-xs`)
added_label
are you sure that it is actually  Issue Comment?

1.4. When I added `bounty-xs` label to [issue](https://github.com/NewOrGanizationSOB/RepoSOB/issues/60) with deployed contract, I got an error.
From webhook:
* [request payload](https://github.com/status-im/autobounty/files/1677356/payload.txt)

* [response](https://github.com/status-im/autobounty/files/1677362/resp.txt)

* [console log output](https://github.com/status-im/autobounty/files/1677370/log.log)

1.5. I couldn't find any log in the system except [console log output](https://github.com/status-im/autobounty/files/1677370/log.log)

## 2. Security

It is no secure to save webhook without `Secret` because anybody who have access to real Payload URL can add it to own repo.

## 3. Functionality questions
cc @andytudhope 

3.1. How token values calculates in USD and should it be updated (because between adding label and actually funding we have `delayInMiliSeconds` time gap)?

what is expected bot behaviour if:

* 3.2. somebody delete SOB bot comment with contract address?

* 3.3. label `bounty-xs` was added BEFORE  SOB bot comment or while it was in "Deploying" state?

* 3.4. if we add `bounty-xs`, than after `delayInMiliSeconds` is run out, delete it and add `bounty-s` label?

3.5. should it be option to configure decimal values in `priceHour` and `BOUNTY_LABELS`?

**Comment by andytudhope** [comment]
@jomsdev please look at the comment above. Also, there are still some TODOs left in the tests, which would be nice to see completed. Let me answer what I can so far though @churik:

1.2. `eth.sendTransaction` happens [here](https://github.com/jomsdev/autobounty/blob/development/index.js#L76) and, the way this is set up would require running a node on the same docker machine with the `sourceAddress` unlocked. If I am right about this, it is not ideal - we would prefer to sign the transaction locally in the bot (which means keeping the private key in a secure ENV variable) and then submit the signed transaction blob to an API endpoint like the ones provided by infura (seeing as it is being used elsewhere anyway). The best library I know for this is likely [ethers.js](https://github.com/L4ventures/ethers.js).

1.3. I think you have the flow wrong @churik. Generally the `bounty-size` label is assigned at the `bounty-awaiting-approval` stage and so already exists when you add the actual `bounty` label that causes the bot to deploy a contract and comment on the issue. So, I think this should be fine as is.

2.1. Agree that we should have a webhook secret, absolutely.

3.1. USD amounts are calculated [here](https://github.com/jomsdev/autobounty/blob/development/bot/prices.js#L44) and seem good to.

3.2. That is a problem with SOB in general and I'm not sure we will sort it out here.

3.3. My understanding is that the label SHOULD be added before the bounty is deployed. While in the `Deploying` state should not be an issue - the bot can only get the info it needs to work once the contract has been deployed and the comment updated with the relevant info about the address etc.

3.4. That is the organisation's problem -  the `delayInMilliseconds` is intended to make sure that, if they mistakenly assign the incorrect label by clicking the wrong thing, they have at least some time to correct. I am happy with it for v1 of this bot.

**Comment by churik** [comment]
Thanks @andytudhope 
1.3 Just to clarify:

> I think you have the flow wrong @churik. Generally the bounty-size label is assigned at the bounty-awaiting-approval stage and so already exists when you add the actual bounty label that causes the bot to deploy a contract and comment on the issue. So, I think this should be fine as is.
Workflow:
create an issue

1. add `bounty-awaiting-approval` label
2. add `bounty-size` label - when `delayInMilliseconds` is run out that is not possible to change `bounty-size` label and fund it, so changes in labels doesn't affect to actually funding
3. add `bounty` label
4. when contract will be successfully deployed it should be funded accordingly to LAST label which is set before `delayInMilliseconds` is run out. 

Is everything correct?

**Comment by positiveblue** [comment]
Hello,

Thanks @andytudhope for answering some of the @churik questions.

Related to the first comment:

1.0  The `config.txt` has not a .js extension so it does not work. The motive is that we merge the user config file with our default.js file and we get it from `require("./" + (process.env.NODE_ENV || "development") + ".js");`

1.1 The `githubUsername` should match the name of the account/bot who posts the information of the smart contracts that have to be found. It has to follow the current format of `status-open-bounty`. The `sourceAddress` is the account from which the bounties will be funded.

1.2 @andytudhope is right on the post before this. We used `eth.sendTransaction` because it was there before we arrived. There shouldn't be any problem changing it to [ethers.js](https://github.com/L4ventures/ethers.js)

1.4 Did the `log` folder exists before executing the server?

2.0 No problem adding the secret to the github webhook and checking that all the requests come form github.


Related to autobounty bot flow:

The bot is waiting for comments in issues of our project. At some point the `githubUser` account will create a comment which contains the address of the smart contract. Is then when the request start the process for funding the issue and call the `delayInMilliseconds`. After that. we will make a request to the Github API getting the bounty labels for that issue and funding the contract with that data.

**Comment by churik** [comment]
1.0 I just link config.txt, but I used file `production.js` with config.txt content. I just cannot insert production.js in GitHub issue because of extension.

1.1 Got it, will fix

1.4 Yep, the `log` folder exists

**Comment by andytudhope** [comment]
@jomsdev that all sounds good to me - are you still working on getting ethers.js in, or is there something else blocking here?

I would expect that `eth_sendTransaction` is the incorrect thing to be using here, seeing as that will require an account to sign (or a full node synced with an unlocked account, which is not ideal. Have you figured out how ether.js works with keys and signing and seen how to sign a transaction blob and then submit that to the network?

**Comment by kafkasl** [comment]
@andytudhope we are working on getting ethers.js, we were a bit busy due to the fosdem 2018. We'll update you shortly.

**Comment by andytudhope** [comment]
@jomsdev @kafkasl Hope you guys enjoyed FOSDEM! Very jealous 😉 Would you mind giving me an update here when you can?

**Comment by positiveblue** [comment]
@andytudhope I am working on the last details of this PR but I have some doubts about `ethers.js`

Following the example from the documentation I got this function

```javascript
const ethers = require('ethers');
const Wallet = ethers.Wallet;
const providers = ethers.providers;

const wallet = new Wallet(config.privateKey);
const provider = providers.getDefaultProvider();

const sendTransaction = function (to, amount, gasPrice) {
    const transaction = {
        nonce: 0,
        gasLimit: config.gasLimit,
        gasPrice: gasPrice,
        to: to,
        value: amount,
        data: "0x",
        // This ensures the transaction cannot be replayed on different networks
        chainId: providers.Provider.chainId.homestead
    };
    
    const signedTransaction = wallet.sign(transaction);
    
    return new Promise((resolve, reject) => {
        provider.sendTransaction(signedTransaction)
            .then(function(hash) {
                resolve(hash);
            }).catch(function(err) {
                reject(err);
            });
    });   
}
```

The function creates a wallet using the user's private key, but does this private key contain the public key too? If not, how does the provider validate the transaction after we signed it?

**Comment by andytudhope** [comment]
EDIT: from my reading of the documentation [here](https://docs.ethers.io/ethers.js/html/api-wallet.html#creating-instances) for instance, you just need to do `wallet.address` to get what you're looking for.

I'd be surprised if there were not an easy way within ethers.js to do this - but check here: https://ethereum.stackexchange.com/questions/3542/how-are-ethereum-addresses-generated and here https://github.com/ethers-io/ethers.js/ and here (though I doubt it necessary to go this far down): https://github.com/ethereumjs/ethereumjs-util/blob/d03528e7da885539cad141c99ea5b88829f73e72/index.js#L302 @jomsdev

**Comment by positiveblue** [comment]
@andytudhope  so yes, the function that I posted before is what we needed here. 

I just asked to be sure because one of the main complications here is how to test not the bot behaviour but that the transactions are really done.

**Comment by kafkasl** [comment]
@andytudhope we have implemented the verification of the signature coming from github's secret and mounted a folder to see the logs outside the docker (for issue 1.4). 

To test the transactions we've decided to create a test account with funds in ropsten testnet to make sure transactions are correct. I'll let you now when this tests are done.

File: bot/github.js
Comment (pedropombeiro) [inline bot/github.js]: This could be a const

File: bot/github.js
Comment (pedropombeiro) [inline bot/github.js]: Same here

File: bot/index.js
Comment (pedropombeiro) [inline bot/index.js]: Same here

File: bot/index.js
Comment (pedropombeiro) [inline bot/index.js]: And here :-)

File: bot/index.js
Comment (pedropombeiro) [inline bot/index.js]: ...

File: readme.md
Comment (pedropombeiro) [inline readme.md]: Typo in payload

File: bot/index.js
Comment (pedropombeiro) [inline bot/index.js]: const

File: bot/index.js
Comment (pedropombeiro) [inline bot/index.js]: const

**Review by pedropombeiro — CHANGES_REQUESTED** [review]
Not a real review, just noticed a few places that could be changed to `const`s. Looking forward to having this in production!

**Comment by kafkasl** [comment]
thanks for the review, we've finally been able to mine some ether on ropsten testnet to fully test the transaction part, so we hope rius is gets to production soon enough!

**Comment by andytudhope** [comment]
I'm sorry @kafkasl - you should have said you were waiting to mine, you could have always just used this Ropsten faucet 😉Over-communicate with us: we can help!! https://faucet.metamask.io/

**Comment by kafkasl** [comment]
@andytudhope we tried some faucets and they didnt work (always empty). We've mined some ether on ropsten and tested the transactions on that net. Major changes:
* docker is mounted with a shared log folder so the container logs are visible from outside
* transactions were tested in ropsten network and worked
* cleanup and added some const as @PombeirP suggested

Let us know if you have any doubt or requirement before merging.

**Comment by andytudhope** [comment]
Hey @kafkasl! Thank you 🎉 I am testing now. Out of interest, where are you running your docker containers (which service provider) and could you link to the issue where it was successfully tested?

**Comment by kafkasl** [comment]
@andytudhope we are running them in a aws instance. I don't understand what di you mean by linking to the issue, we set up a testing account on ropsten, triggered the bot from a test repo webhook, and checked that the funds are transferred. I can give more details and visual confirmation if you want

**Comment by andytudhope** [comment]
Yeah @kafkasl - a little more detail, visual confirmation, and a link to the test repo you were using is what I am looking for 👍 I am also running it on an AWS instance, with an elastic IP so that it is web accessible and not having any luck.

I also wonder why you were using `jomsdev` as the `githubUsername` and not `status-open-bounty` as that is the name of the bot making the relevant comment, right?

File: bot/github.js
Comment (andytudhope) [inline bot/github.js]: Would this not need to be changed for different organizations?

Comment (kafkasl) [reply bot/github.js]: @andytudhope 'User-agent' is obtained from the configuration file (it was hardcoded in this version I am pushing the update with the fix) the hostname is common (for all using github as end-point).

**Comment by kafkasl** [comment]
@andytudhope, @jomsdev configured the machine, I can give you access if you want. The `jomsdev` username was used for testing because we posted issues to our test repo with our own usernames `jomsdev` or `kafkasl` because we can not post issues with `status-open-autobounty` name. 

I am attaching the result of posting an issue to our test repo and the according transaction. (the repo is private, I will make it public shortly because I have not the permissions yet)

log
github
etherscan

**Comment by andytudhope** [comment]
OK @kafkasl - no need to give me access. Thanks for answering my review comment too.

I have a machine running with docker and the webhooks are getting through fine. However, there is an additional bit of complexity here not present in your test of just posting a contract address. Unfortunately, when `status-open-bounty` posts a comment, it first is about `deploying contract` and only updates a little later with the actual address to be funded.

This is what that webhook looks like (the relevant part at the top at least):

```
{
  "action": "edited",
  "changes": {
    "body": {
      "from": "![Contract deploying](https://openbounty.status.im:444/img/deploying_contract.png)\nTransaction: [0x3c7e100546c004c78c8c8cb83c09555efe8b7d56d5dfed86de1ee2a3b84b8e50](https://ropsten.etherscan.io/tx/0x3c7e100546c004c78c8c8cb83c09555efe8b7d56d5dfed86de1ee2a3b84b8e50)\n"
    }
  },
  "issue": {
    "url": "https://api.github.com/repos/andytudhope/react-timer/issues/4",
    . . . [a whole bunch more info here, omitted for brevity]
  },
   "comment": {
    "url": "https://api.github.com/repos/andytudhope/react-timer/issues/comments/369581005",
    "html_url": "https://github.com/andytudhope/react-timer/issues/4#issuecomment-369581005",
    "issue_url": "https://api.github.com/repos/andytudhope/react-timer/issues/4",
    "id": 369581005,
    "user": {
      "login": "status-open-bounty",
      "id": 33065743,
      "avatar_url": "https://avatars3.githubusercontent.com/u/33065743?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/status-open-bounty",
      "html_url": "https://github.com/status-open-bounty",
      "followers_url": "https://api.github.com/users/status-open-bounty/followers",
      "following_url": "https://api.github.com/users/status-open-bounty/following{/other_user}",
      "gists_url": "https://api.github.com/users/status-open-bounty/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/status-open-bounty/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/status-open-bounty/subscriptions",
      "organizations_url": "https://api.github.com/users/status-open-bounty/orgs",
      "repos_url": "https://api.github.com/users/status-open-bounty/repos",
      "events_url": "https://api.github.com/users/status-open-bounty/events{/privacy}",
      "received_events_url": "https://api.github.com/users/status-open-bounty/received_events",
      "type": "User",
      "site_admin": false
    },
    "created_at": "2018-03-01T12:47:18Z",
    "updated_at": "2018-03-01T12:49:17Z",
    "author_association": "NONE",
    "body": "Current balance: 0.000000 ETH\nContract address: [0xb9d815634d068875191a161a71c1a8532b1d0679](https://ropsten.etherscan.io/address/0xb9d815634d068875191a161a71c1a8532b1d0679)\n![QR Code](https://openbounty.status.im:444/qr/andytudhope/react-timer/bounty/4/b5989f87a47aad94004bee8b726155d7b5c2365b12403ac73777d2701edeb840/qr.png)\nNetwork: Testnet (Ropsten)\nTo claim this bounty sign up at [https://openbounty.status.im:444](https://openbounty.status.im:444) and make sure to update your Ethereum address in `My Payment Details` so that the bounty is correctly allocated.\nTo fund it, send test ETH or test ERC20/ERC223 tokens to the contract address."
  },
  . . .
}
```

To be clear then, there are 2 payloads sent to the webhook each time a bounty is created. The first starts with `"action": "created",` and can be safely ignored as it is about deploying the contract. The second starts, as above, with `"action": "edited"` and has the relevant contract address.

So you want to look for the `edited` action and use the `comment.body` to access the address that needs to be funded. I think we may need just a few more changes. We are so close though! And I really appreciate your guys' work. This has turned into a lot more effort than we initially anticipated, so I will increase the size of the bounty to better reflect all the work and time you guys have put into this. Thank you again! ❤️

**Comment by andytudhope** [comment]
Also @kafkasl you absolutely can deploy bounties yourself to test more thoroughly. Simply install [this test GH app](https://github.com/apps/status-open-bounty-app-test) in the relevant repo (it'll need to be a public one though), add a `bounty` label and fire away 👍

**Comment by andytudhope** [comment]
Actually @kafkasl couldn't this be achieved with minimal effort by calling `hasAddress` before `getAddress` in the `index.js` file in the root directory?

File: bot/index.js
Comment (andytudhope) [inline bot/index.js]: `if (req.body.action !== 'edited' || !req.body.hasOwnProperty('comment')) {`

Would be my suggestion

File: bot/index.js
Comment (andytudhope) [inline bot/index.js]: `return commentBody.substring(commentBody.search("Contract address:") + 19, commentBody.search("Contract address:") + 61)` because what actually gets posted in that webhook seems to be

`"body": "Current balance: 0.000000 ETH\nContract address: [0xdff55bc9962ce948597eeeaa1339ebb7c70f6a7c](https://ropsten.etherscan.io/address/0xdff55bc9962ce948597eeeaa1339ebb7c70f6a7c)\nQR Code\nNetwork: Testnet (Ropsten)\nTo claim this bounty sign up at [https://openbounty.status.im:444](https://openbounty.status.im:444) and make sure to update your Ethereum address in `My Payment Details` so that the bounty is correctly allocated.\nTo fund it, send test ETH or test ERC20/ERC223 tokens to the contract address."`

**Comment by kafkasl** [comment]
@andytudhope thanks for appreciating the work and increasing the bounty ;). About the `action: edited` we were actually filtering the requests with `action !== "created"` so the result was the same, however I'm pushing the change to filter by `action === "edited"` which is more precise.

File: config/development.js
Comment (andytudhope) [inline config/development.js]: what abut `realTransaction` here? Can we just have one config file as well, rather than default and development, as the different forms in there are confusing.

**Comment by kafkasl** [comment]
@andytudhope I tested the new filtering of the issue the old way and it works. I wanted to try also with GH app but I'm not sure how to make it work after installing it.

**Review by andytudhope — COMMENTED** [review]
With these changes I can get it to run and the webhook gets a `200` response as expected. However, it doesn't seem to fund the issue...

I have an account on ropsten with both ETH and `STT` in it and have included `STT` in the config file, so am not sure where I am going wrong...

**Comment by andytudhope** [comment]
@kafkasl all you need to do, after installing the GH app on the repo you want to issue a bounty on, is create a `bounty` label and then use that `bounty` label on any issue - the bot should then come in an comment automatically without you needing to do anything else.

**Comment by andytudhope** [comment]
Also, just worked out why using `STT` would be an issue - it won't get fetched when checking token price. Will get some SNT on ropsten somehow and then test again.

File: config/development.js
Comment (kafkasl) [reply config/development.js]: yes, I merged them and now the default contains all the development values

File: bot/index.js
Comment (kafkasl) [reply bot/index.js]: @andytudhope I'm not sure of the issue here, we copied the examples format of status-open-bounty. Are you saying that they were wrong and instead we should parse another format? 

by the way, there is a ./log folder were you run the bot container. Could you post the contents when you get a 200 but the transaction is not done?

**Comment by kafkasl** [comment]
@andytudhope I installed the GH app on this repo and commented, https://github.com/kafkasl/my-github-bot/issues but it does not appear to be working or maybe I'm missing something.

**Comment by andytudhope** [comment]
@kafkasl sorry, I missed one step. Please also log into http://openbounty.status.im:444 (the test environment) and update your address in the payments details dropdown there as currently the error is `Unable to deploy bounty contract because repo owner has no Ethereum address`.

**Comment by kafkasl** [comment]
@andytudhope where are the error logs? I added my ethereum address but it is still not working.

**Comment by andytudhope** [comment]
sorry @kafkasl the error logs are on the machine itself. Vitaliy has fixed it now, so you should be good to go again there, apologies for the hold up!

**Comment by kafkasl** [comment]
@andytudhope tried again to no avail. https://github.com/kafkasl/my-github-bot/issues/12 , are you sure I'm not missing something I need to do?

**Comment by andytudhope** [comment]
Yes @kafkasl - you need to put just a `bounty` label on it, and then a size label like you have already, i.e. `bounty-s`. i.e. you'll need 2 labels.

**Comment by kafkasl** [comment]
@andytudhope I can't make it work. I just created an issue an put both labels bounty and bounty-s but nothing happens.

**Comment by andytudhope** [comment]
I'm not sure what is going wrong @kafkasl? Can you please remove the GH app (which should be [this one](https://github.com/apps/status-open-bounty-app-test)), make sure you have logged into https://openbounty.status.im:444  and then re-add it? I'm sorry this is dragging on, we've had problems with our test environment recently 😞

**Comment by kafkasl** [comment]
@andytudhope ok. I finally got it working. The contract address length was a bit off and the condition is now working for edited, not created. Everything tested and working now. 
Issue: https://github.com/kafkasl/my-github-bot/issues/22
Funding transaction: 0x761f8cf6afe5d82a3b18e8a07e811950eeb9a822c8439372e5cb7b1a2d2c93f2
confirmation

**Comment by andytudhope** [comment]
So, I tried all weekend to get this working and had no success, I'm really sorry. I think the biggest issue is that you're not handling token sends correctly (which cannot just be done by `eth.sendTransaction` unfortunately). You need to target the `transfer` function on the smart contract that controls that particular token.

@PombeirP is there any way that you could help with this this week. It has dragged on forever and we need to get it done so that we can award bounties here for the work done and move on. Will fill you in in Slack if you do have a few hours to help with this.

**Pull request merged**
